// @allurereport/service/dist/service.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureServiceClient_instances, _AllureServiceClient_client, _AllureServiceClient_url, _AllureServiceClient_pollingDelay, _AllureServiceClient_getClientUrl;
import { readFile } from "node:fs/promises";
import { join as joinPosix } from "node:path/posix";
import open from "open";
import { createServiceHttpClient } from "./utils/http.js";
import { decryptExchangeToken, deleteAccessToken, writeAccessToken, writeExchangeToken } from "./utils/token.js";
const ASSET_MAX_FILE_SIZE = 200 * 1024 * 1024;
export class AllureServiceClient {
    constructor(config) {
        _AllureServiceClient_instances.add(this);
        this.config = config;
        _AllureServiceClient_client.set(this, void 0);
        _AllureServiceClient_url.set(this, void 0);
        _AllureServiceClient_pollingDelay.set(this, void 0);
        if (!config.url) {
            throw new Error("Allure service URL is required!");
        }
        __classPrivateFieldSet(this, _AllureServiceClient_url, config.url, "f");
        __classPrivateFieldSet(this, _AllureServiceClient_client, createServiceHttpClient(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), config?.accessToken), "f");
        __classPrivateFieldSet(this, _AllureServiceClient_pollingDelay, config?.pollingDelay ?? 2500, "f");
        this.currentProjectUuid = config?.project;
    }
    setProject(project) {
        this.currentProjectUuid = project;
    }
    async login() {
        const clientUrl = await __classPrivateFieldGet(this, _AllureServiceClient_instances, "m", _AllureServiceClient_getClientUrl).call(this);
        const exchangeToken = await writeExchangeToken();
        const connectUrl = new URL("/connect", clientUrl);
        connectUrl.searchParams.set("token", decryptExchangeToken(exchangeToken));
        await open(connectUrl.toString());
        let currentExchangeAttemptTimeout;
        return await new Promise((res) => {
            const makeExchangeAttempt = () => {
                return globalThis.setTimeout(async () => {
                    const token = decryptExchangeToken(exchangeToken);
                    const { accessToken } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/auth/exchange", {
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: {
                            token,
                        },
                    });
                    if (!accessToken) {
                        globalThis.clearTimeout(currentExchangeAttemptTimeout);
                        currentExchangeAttemptTimeout = makeExchangeAttempt();
                        return;
                    }
                    await writeAccessToken(accessToken);
                    return res(accessToken);
                }, __classPrivateFieldGet(this, _AllureServiceClient_pollingDelay, "f"));
            };
            currentExchangeAttemptTimeout = makeExchangeAttempt();
        });
    }
    async logout() {
        await deleteAccessToken();
    }
    async profile() {
        const { user } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/user/profile");
        return user;
    }
    async projects() {
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/projects");
    }
    async project(uuid) {
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get(`/projects/${uuid}`);
    }
    async createProject(payload) {
        const { project } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/projects", {
            body: payload,
        });
        return project;
    }
    async deleteProject(payload) {
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").delete(`/projects/${payload.id}`);
    }
    async downloadHistory(branch) {
        if (!this.currentProjectUuid) {
            throw new Error("Project is not set");
        }
        const { history } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get(`/projects/${this.currentProjectUuid}/${branch}/history`);
        return history;
    }
    async createReport(payload) {
        const { reportName, reportUuid, branch } = payload;
        if (!this.currentProjectUuid) {
            throw new Error("Project is not set");
        }
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/reports", {
            body: {
                projectUuid: this.currentProjectUuid,
                reportName,
                reportUuid,
                branch,
            },
        });
    }
    async completeReport(payload) {
        const { reportUuid, historyPoint } = payload;
        if (!this.currentProjectUuid) {
            throw new Error("Project is not set");
        }
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/complete`, {
            body: {
                historyPoint,
            },
        });
    }
    async deleteReport(payload) {
        const { reportUuid, pluginId = "" } = payload;
        if (!this.currentProjectUuid) {
            throw new Error("Project is not set");
        }
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/delete`, {
            body: {
                pluginId,
            },
        });
    }
    async addReportAsset(payload) {
        const { filename, file, filepath } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Asset size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", filename);
        form.set("file", content);
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/assets/upload", {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });
    }
    async addReportFile(payload) {
        const { reportUuid, filename, file, filepath, pluginId } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Report file size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", pluginId ? joinPosix(pluginId, filename) : filename);
        form.set("file", content);
        await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/upload`, {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });
        return joinPosix(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), reportUuid, filename);
    }
}
_AllureServiceClient_client = new WeakMap(), _AllureServiceClient_url = new WeakMap(), _AllureServiceClient_pollingDelay = new WeakMap(), _AllureServiceClient_instances = new WeakSet(), _AllureServiceClient_getClientUrl = async function _AllureServiceClient_getClientUrl() {
    const { url } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/info");
    return url;
};
