// @allurereport/service/dist/utils/http.js
import axios, { isAxiosError } from "axios";
import { DEFAULT_HISTORY_SERVICE_URL } from "../model.js";
import { readAccessToken } from "./token.js";
export class KnownError extends Error {
    constructor(message, status) {
        super(message);
        this.name = "KnownError";
        this.status = status;
    }
}
export class UnknownError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = "UnknownError";
        this.stack = stack;
    }
}
export const createServiceHttpClient = (historyServiceURL = DEFAULT_HISTORY_SERVICE_URL, accessToken) => {
    const client = axios.create({
        baseURL: historyServiceURL,
        withCredentials: true,
        validateStatus: (status) => status < 400,
    });
    const sendRequest = (method) => async (endpoint, payload) => {
        const actualAccessToken = accessToken || (await readAccessToken());
        const headers = {
            ...(payload?.headers ?? {}),
        };
        if (actualAccessToken) {
            headers.Authorization = `Bearer ${actualAccessToken}`;
        }
        try {
            let res;
            if (payload?.body) {
                res = await client[method](endpoint, payload.body, {
                    ...payload,
                    headers,
                });
            }
            else {
                res = await client[method](endpoint, {
                    ...payload,
                    headers,
                });
            }
            return res.data;
        }
        catch (err) {
            const axiosError = isAxiosError(err);
            if (!axiosError) {
                throw err;
            }
            const { status = 500 } = err.response ?? {};
            const errorMessage = err.response?.data?.error || err.response?.data || err.message;
            if (status < 500) {
                throw new KnownError(errorMessage, status);
            }
            throw new UnknownError(errorMessage, err.stack);
        }
    };
    return {
        get: sendRequest("get"),
        post: sendRequest("post"),
        put: sendRequest("put"),
        delete: sendRequest("delete"),
    };
};
